### ISSUE TITLE
Make bucket based remotes "rootless"

### ISSUE THREAD

**Issue by ncw** [rclone/rclone#3421]
Currently bucket based remotes such as

- [x] azureblob
- [x] b2
- [x] gcs
- [x] qingstor
- [x] s3
- [x] swift

Have various limitations due to only being able to be instantiated with a bucket.
- mount/serve doesn't work on the bare `remote:`
- there are work arounds in --rc-serve for bucket based remotes
- there are work arounds in the web gui

To work around this bucket based remotes need to work from the root for any path.  This is a fairly simple change and a POC has been developed for the swift backend.

As part of this
- remove -subdir flag and put an explicit flag in fstests
    - NB -subdir isn't used for backend tests!
- test for root listing and operations

**Comment by yparitcher** [comment]
When doing this is it possible to also allow for a remote to be rooted at a specific bucket. fo example `remote:` would equal `B2:Bucket`. this is very useful when using keys limited to one bucket.

maybe this can be an option in the config?

thanks

**Comment by ncw** [comment]
> When doing this is it possible to also allow for a remote to be rooted at a specific bucket. fo example `remote:` would equal `B2:Bucket`. this is very useful when using keys limited to one bucket.
> 
> maybe this can be an option in the config?

You can do this with the [alias backend](https://rclone.org/alias/) - is that good enough?

**Comment by yparitcher** [comment]
> You can do this with the alias backend <https://rclone.org/alias/> - 
> is that good enough?
perfect

**Comment by yparitcher** [comment]
from reading the B2 changes

should 
https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L755-L757
and
https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L790-L791

have consistent calls? if `bucket == ""`, `directory` will also be empty.

---

https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L813
only marks one bucket present
something like:
```
	if bucket == "" {
		entries, err := f.listBuckets("")
		if err != nil {
			return err
		}
	} else {
		entries := bucket
	}
	for _, entry := range entries {
		err = list.Add(entry)
		if err != nil {
			return err
		}
		bucket := entry.Remote()
		err = listR(bucket, directory, f.rootDirectory, f.rootBucket == "")
		if err != nil {
			return err
		}
		// bucket must be present if listing succeeded
		f.cache.MarkOK(bucket)
	}
```
might be simpler.

these probably apply to all the remotes

**Comment by yparitcher** [comment]
https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L1192

is testing for the root not the dst bucket.

maybe
```go
	err := f.Mkdir(ctx, dstBucket)
```
---
not sure about this one 
https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L1673

**Comment by ncw** [comment]
> from reading the B2 changes
> 
> should
> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L755-L757
> 
> 
> and
> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L790-L791
> 
> have consistent calls? if `bucket == ""`, `directory` will also be empty.

Yes.  The fact that listBuckets() takes an argument at all is a bit of an anachronism.

We should probably remove the argument from listBuckets and check that directory == "" in List

> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L813
> 
> 
> only marks one bucket present
> something like:
> ```go
> 	if bucket == "" {
> 		entries, err := f.listBuckets("")
> 		if err != nil {
> 			return err
> 		}
> 	} else {
> 		entries := bucket
> 	}
> 	for _, entry := range entries {
> 		err = list.Add(entry)
> 		if err != nil {
> 			return err
> 		}
> 		bucket := entry.Remote()
> 		err = listR(bucket, directory, f.rootDirectory, f.rootBucket == "")
> 		if err != nil {
> 			return err
> 		}
> 		// bucket must be present if listing succeeded
> 		f.cache.MarkOK(bucket)
> 	}
> ```
> 
> might be simpler.

Yes that would work.  I think doing `rclone ls s3:` to list **all** your buckets is probably not too common though.

> these probably apply to all the remotes

Indeed :-) 

I'd like to have a go at factoring out ListR into the bucket package - that should be possible I think...

> `Mkdir(ctx, "")` does nothing
> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L1192
> 
> is testing for the root not the dst bucket.

That should be making the destination bucket if it doesn't exist which it will if the destination is `b2:bucket/path`

I see what you are saying though, if the destination is `b2:` then it will do nothing

> maybe
> 
> ```go
> 	err := f.Mkdir(ctx, dstBucket)
> ```

Actually I think something like `f.Mkdir(ctx, path.Dir(remote))` is correct

> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/backend/b2/b2.go#L1673-L1679
> 
> ```go
> 	bucket, bucketPath := o.split()
> 
> 	err = o.fs.Mkdir(ctx, bucket)
> 	if err != nil {
> 		return err
> 	}
> 	size := src.Size()
> ```

I don't think that is correct.  Remember Mkdir works relative to the root.  I think the same fix as above would work `Mkdir(path.Dir(o.remote))`, or maybe factoring `Mkdir` into a `createBucket` might be clearer so that could be called `f.createBucket(bucket)`

> https://github.com/rclone/rclone/blob/6ca00c21a4e3478c2f639e9e0a34adf0e7f4476e/lib/bucket/bucket.go#L81-L89
> 
> 
> should not need the mutex and can be switched
> ```go
> 	// if we are at the root, then it is OK
> 	if bucket == "" {
> 		return nil
> 	}
> 
> 	c.createMu.Lock()
> 	defer c.createMu.Unlock()
> 	c.mu.Lock()
> 	defer c.mu.Unlock()
> ```

Noted.

Thank you very much for the code review :-)

I'll do a bit of fixing up.

**Comment by yparitcher** [comment]
> Yes. The fact that listBuckets() takes an argument at all is a bit of an anachronism.

> We should probably remove the argument from listBuckets and check that directory == "" in List

good idea
> Actually I think something like `f.Mkdir(ctx, path.Dir(remote))` is correct

> I don't think that is correct. Remember Mkdir works relative to the root. I think the same fix as above would work `Mkdir(path.Dir(o.remote))`, or maybe factoring `Mkdir` into a `createBucket` might be clearer so that could be called `f.createBucket(bucket)`

right now Mkdir does the same as a createBucket would do (i mentioned this in #3453). since a bucket is always top level to the root, and Mkdir discards the rest of the path, calling Mkdir on the bucket or path should be the same. 
mkdir calls f.split() and just uses the bucket part.
i agree that for now it should be renamed/factored into a createBucket to avoid confusion
this is for B2, it might be different in the other remotes, idk

**Comment by ncw** [comment]
> right now Mkdir does the same as a createBucket would do (i mentioned this in #3453). since a bucket is always top level to the root, and Mkdir discards the rest of the path, calling Mkdir on the bucket or path should be the same.

Oh, I see what you mean... Not sure I like that idea but I understand now :-)

**Comment by yparitcher** [comment]
When reading the remotes for #3453 i saw that in these commits you changed the signature of the `list` function from 
```go
list(ctx context.Context, dir string, ...)
```
to 
```go
list(ctx context.Context, bucket, directory, prefix string, ...)  
```
or similar.

for all the remotes i saw so far the `prefix` was set to `rootDirectory` (or for a file was `""`). 

`rootDirectory` is set by `newFs` and therefore should always (if i am reading the code right) be the same as `directory` and by changing from one `dir` string to 2 (`directory, prefix`) a lot of useless complexity was added. 

i think this should be refactored on all the BBR to only have one: either `directory` or `prefix` and streamline the list command.

i think this came about as the prefix used to be set from `dir` leading to this confusion.
please let me know if i am wrong

**Comment by ncw** [comment]
> When reading the remotes for #3453 i saw that in these commits you changed the signature of the `list` function from
> 
> ```go
> list(ctx context.Context, dir string, ...)
> ```
> 
> to
> 
> ```go
> list(ctx context.Context, bucket, directory, prefix string, ...)  
> ```
> 
> or similar.

I wanted the `list()` commands to be completely and utterly explicit,  So (bucket, directory) is what we are listing, `prefix` what we are removing off the names received and the flag `addBucket` controls whether we add the bucket name on to them.

> for all the remotes i saw so far the `prefix` was set to `rootDirectory` (or for a file was `""`).

This may be true :-)
 
> `rootDirectory` is set by `newFs` and therefore should always (if i am reading the code right) be the same as `directory` and by changing from one `dir` string to 2 (`directory, prefix`) a lot of useless complexity was added.

I don't think `directory` and `prefix` are the same, for example in `List` `directory` is the path to the subdirectory and `prefix` is the `rootDirectory`.  So if the remote is `swift:bucket/dir` then `rootBucket` is `bucket` and `rootDirectory` is `dir`.  If you do `List("subdir")` then we call `f.list("bucket", "dir/subdir", "dir", false`).

> i think this should be refactored on all the BBR to only have one: either `directory` or `prefix` and streamline the list command.
> 
> i think this came about as the prefix used to be set from `dir` leading to this confusion.
> please let me know if i am wrong

**Comment by ncw** [comment]
This is done and was released in v1.50
